src/ParseFasm.cpp

```diff
@@ -0,0 +1,508 @@
+// Copyright (C) 2020  The SymbiFlow Authors.
+//
+// Use of this source code is governed by a ISC-style
+// license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/ISC
+//
+// SPDX-License-Identifier: ISC
+
+#include "FasmLexer.h"
+#include "FasmParser.h"
+#include "FasmParserVisitor.h"
+#include "antlr4-runtime.h"
+
+/// This code parses FASM and produces a lightweight binary format that
+/// is fast and simple to unpack based on the tag/length/value (TLV)
+/// format.
+///
+/// For example, for a fixed width integer, this can just be:
+///   <tag : 1 byte> <data : 4 bytes>
+/// and for a variable length string:
+///   <tag : 1 byte> <length : 4 bytes> <data : length bytes>
+///
+/// Note that the length itself must be a fixed width value, which does
+/// impose a size limit, but this format is more efficient and easier
+/// to decode than a UTF-8 style variable length encoding where a bit
+/// is reserved per per byte to indicate the end of a variable length
+/// value.
+///
+/// TLVs can be nested, with each level adding 5 bytes of header
+/// overhead. There is a choice to aggregate values under another header
+/// (using withHeader) or not. Even though this encodes redundant size
+/// information, it can make the result easier to parse.
+///
+/// Example of a nested TLV:
+///   <outer tag : 1 byte> <outer length = 5 + nA + 5 + nB : 4 bytes>
+///     <A tag : 1 byte> <A length : 4 bytes> <A data : nA bytes>
+///     <B tag : 1 byte> <B length : 4 bytes> <B data : nB bytes>
+///
+/// Note that there is no need for a closing tag.
+///
+/// For example, the consumer can allocate space for results in larger
+/// chunks, in this case it can preallocate a line at a time, after
+/// reading the first 5 bytes, although this is mostly useful for
+/// consumers that can use the data directly without further
+/// manipulation.
+///
+/// The format used here does not rely on knowing the size of the
+/// entire output, which allows streaming line by line, although
+/// ANTLR does not yet implement an incremental parser.
+///
+/// For a concrete example, see the test case in ParseFasmTests.cpp
+
+/// External C Interface
+///
+/// These functions serialize the FASM parse tree to an easy to parse
+/// tag/length/value binary format, where the tag is one byte and
+/// the length is 4 bytes, in native endianness (typically little.)
+extern "C" {
+void from_string(const char* in, bool hex, void (*ret)(const char* str));
```

Comment:
Does it make sense to change the `from_string` API to take a string length, rather than require it to be null terminated?  Don't feel strongly about this.

Reply:
I don't think so. `ctypes` converts Python byte objects into null terminated strings.